### PR TITLE
Update scribe notes for week 13 (Graph Convolutional Network)

### PR DESCRIPTION
**Pull request by muneebafzal**
In this section, we introduce Graph Convolutional Network (GCN) which is one type of architecture that utilizes the structure of data.  Actually, the concept of GCNs is closely related to self-attention. After understanding the general notation, representation and equations of GCN, we delve into the theory and code of a specific type of GCN known as Residual Gated GCN.

### CODE REVIEW

File: docs/en/week13/13-3.md [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+train_acc: 0.9429, test_acc: 0.8900
+Epoch 36, train_loss: 0.3873, test_loss: 0.3748
+train_acc: 0.9114, test_acc: 0.9200
+Epoch 37, train_loss: 0.3478, test_loss: 0.3489
+train_acc: 0.9429, test_acc: 0.9400
+Epoch 38, train_loss: 0.3266, test_loss: 0.3475
+train_acc: 0.9543, test_acc: 0.9300
+Epoch 39, train_loss: 0.3885, test_loss: 0.3666
+train_acc: 0.9086, test_acc: 0.8700
+```
```
Comment (Atcold) [inline docs/en/week13/13-3.md:510]: Remove all of this. People have access to the notebook.
This is supposed to be a summary of the lecture.

File: docs/en/week13/13-3.md [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+
+for epoch in range(40):
+    
+    start = time.time()
+    train_loss, train_acc = train(model, train_loader, loss)
+    test_loss, test_acc = evaluate(model, test_loader, loss)
+        
+    print(f'Epoch {epoch}, train_loss: {train_loss:.4f}, test_loss: {test_loss:.4f}')
+    print(f'train_acc: {train_acc:.4f}, test_acc: {test_acc:.4f}')
+
```
Comment (Atcold) [inline docs/en/week13/13-3.md:425]: Nothing special here. People **do** have access to the notebook.

File: docs/en/week13/13-3.md [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+            J = loss(batch_scores, batch_labels)
+
+            epoch_test_loss += J.detach().item()
+            epoch_test_acc += accuracy(batch_scores, batch_labels)
+            nb_data += batch_labels.size(0)
+
+        epoch_test_loss /= (iter + 1)
+        epoch_test_acc /= nb_data
+    
+    return epoch_test_loss, epoch_test_acc
```
Comment (Atcold) [inline docs/en/week13/13-3.md:397]: Not interesting.

File: docs/en/week13/13-3.md [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+        optimizer.step()
+        
+        epoch_loss += J.detach().item()
+        epoch_train_acc += accuracy(batch_scores, batch_labels)
+        nb_data += batch_labels.size(0)
+
+    epoch_loss /= (iter + 1)
+    epoch_train_acc /= nb_data
+
+    return epoch_loss, epoch_train_acc
```
Comment (Atcold) [inline docs/en/week13/13-3.md:369]: Not relevant.

File: docs/en/week13/13-3.md [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+  )
+  (MLP_layer): MLP_layer(
+    (FC_layers): ModuleList(
+      (0): Linear(in_features=100, out_features=100, bias=True)
+      (1): Linear(in_features=100, out_features=100, bias=True)
+      (2): Linear(in_features=100, out_features=8, bias=True)
+    )
+  )
+)
+```
```
Comment (Atcold) [inline docs/en/week13/13-3.md:339]: Way too much code!

File: docs/en/week13/13-3.md [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+        
+    def forward(self, x):
+        y = x
+        for l in range(self.L):
+            y = self.FC_layers[l](y)
+            y = torch.relu(y)
+        y = self.FC_layers[self.L](y)
+        return y
+
+```
```
Comment (Atcold) [inline docs/en/week13/13-3.md:264]: We have already covered MLPs.

File: docs/en/week13/13-3.md [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+        e = self.bn_node_e(e) # batch normalization  
+        
+        h = torch.relu(h) # non-linear activation
+        e = torch.relu(e) # non-linear activation
+        
+        h = h_in + h # residual connection
+        e = e_in + e # residual connection
+        
+        return h, e
+```
```
Comment (Atcold) [inline docs/en/week13/13-3.md:242]: Break down each method, and explain it separately.

File: docs/en/week13/13-3.md (hunk 5 above)
Comment (Atcold) [reply docs/en/week13/13-3.md:339]: Maybe just keep lines 309 and 313?

File: docs/en/week13/13-3.md (hunk 7 above)
Comment (mok232) [reply docs/en/week13/13-3.md]: Done

File: docs/en/week13/13-3.md (hunk 6 above)
Comment (muneebafzal) [reply docs/en/week13/13-3.md]: Cleaned the extra code

File: docs/en/week13/13-3.md (hunk 5 above)
Comment (muneebafzal) [reply docs/en/week13/13-3.md]: removed extra code

File: docs/en/week13/13-3.md (hunk 4 above)
Comment (muneebafzal) [reply docs/en/week13/13-3.md]: removed

File: docs/en/week13/13-3.md (hunk 3 above)
Comment (muneebafzal) [reply docs/en/week13/13-3.md]: removed

File: docs/en/week13/13-3.md (hunk 2 above)
Comment (muneebafzal) [reply docs/en/week13/13-3.md]: removed

File: docs/en/week13/13-3.md (hunk 1 above)
Comment (muneebafzal) [reply docs/en/week13/13-3.md]: Done, extra code blocks have been cleaned

File: docs/en/week13/13-3.md [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+In this graph, vertices are connected with directed edges.
+
+We represent these directed edges with adjacency vector $\boldsymbol{a}$, where each element $\alpha_{j}$ is set to $1$ if there is a directed edge from $v_{j}$ to $v$.
+
+$$
+\alpha_{j} \stackrel{\tiny \downarrow}{=} 1 \Leftrightarrow v_{j} \rightarrow v
+\tag{Eq. 1}
+$$
+
+The degree (number of incoming edges) $d$ is defined as the norm of this adjacency vector, i.e. $\|\|\boldsymbol{a}\|\|_{1} $, which is the number of ones in the vector $\boldsymbol{a}$.
```
Comment (Atcold) [inline docs/en/week13/13-3.md:40]: The norm is obtained using `\Vert`, the absolute value with `\vert`, and the conditional bar with `\mid`.
Please, remove any `||`, because they are interpreted as a Markdown table.

**Comment by Atcold** [comment]
As I've said earlier, there's no summary in `13.md` and you haven't updated the `_config.yml` file.
Let me see if I can quickly hack your fork myself…

**Comment by Atcold** [comment]
Okay, fixed everything myself.
Good write-up! It's really really pretty! Thanks!

**Comment by muneebafzal** [comment]
Thank you so much, Alf!! 👍 👍

**Comment by mok232** [comment]
@Atcold Sorry, it completely skipped my mind earlier. There seems to be a discrepancy between the equations in the updated slides and the notebook. Specifically, in these 2 equations, the superscript 'h' seems to be missing in the equations on the slides. For reference, I have used the equations from the notebook in the notes.

image

image

**Comment by Atcold** [comment]
The equations on the notebook are the actual mistake, haha.
I spent 1 hour to fix the slides, haha, and then failed to port the equations to the notebook.
If you check the implementation, you'll see the slides are describing it correctly.
Please, update notebook (if it's okay with you) and write up.

Thanks so much for looking into this!

**Comment by mok232** [comment]
Haha, I see. Yeah, no worries. I created a PR to fix the equations in the write up and notebook: https://github.com/Atcold/pytorch-Deep-Learning/pull/311

Hopefully, I managed to correct it in all the places. Now, I finally see why you created the separate equation for e<sub>j</sub> in the slides. I guess the confusion was that the equations were not being computed in the same order in the code as presented in the slides. For instance, e<sub>j</sub> is computed before h since the `message_func` function computes the edge representations before the node representations `h` are computed in the `forward`. Defining the separate equation for e<sub>j</sub> definitely helped!
